Fix docker cp dir with hard link

keloyang opened the pull request:
Use docker cp command to copy dir from container, if the path in container is not the same to the path on the host, docker cp will fail, this pr will fix it.

This bug need docker/cli update vendor for docker.

Signed-off-by: yangshukui <yangshukui@huawei.com>

**- What I did**
Adjust targetPath with hard link.

**- How to verify it**

***before this pr
![1](https://user-images.githubusercontent.com/11223367/27851423-31354bda-618d-11e7-9b40-d282b81c2728.JPG)

***after this pr
![2](https://user-images.githubusercontent.com/11223367/27851437-3c2b5214-618d-11e7-8dcf-fd7a840f010e.JPG)

thaJeztah commented on pkg/archive/archive.go:
```diff
@@ -530,7 +530,21 @@ func createTarFile(path, extractDir string, hdr *tar.Header, reader io.Reader, L
+		//hdr.Name = "tmp_/b"
+		//--> lpath = tmp_/a -->targetPath =extractDir + lpath --> ln tmp_/a tmp_/b
+		lpath := hdr.Linkname
+		flinkName := strings.Split(hdr.Linkname, "/")
```
Should this always use `/`, or `os.PathSeparator`? I see other parts of the code use, e.g. `filepath.Join()`, which is platform dependent

thaJeztah commented on pkg/archive/archive.go:
```diff
@@ -530,7 +530,21 @@ func createTarFile(path, extractDir string, hdr *tar.Header, reader io.Reader, L
+			fName := strings.Split(hdr.Name, "/")
+			if len(fName) > 1 {
+				flinkName[0] = fName[0]
+				lpath = strings.Join(flinkName, "/")
```
Same here; `strings.Join()`, or `filepath.Join()`?

keloyang replied:
@thaJeztah, thanks for your advice, I have modified.

thaJeztah commented on pkg/archive/archive.go:
```diff
@@ -530,7 +530,21 @@ func createTarFile(path, extractDir string, hdr *tar.Header, reader io.Reader, L
+		if len(flinkName) > 1 {
+			fName := strings.Split(hdr.Name, "/")
+			if len(fName) > 1 {
+				flinkName[0] = fName[0]
```
Trying to understand the fix, so bear with me 😊 - does this always only have to replace the first path element? Would there be cases where more elements need to be copied? would something like [`filepath.Rel()`](https://golang.org/pkg/path/filepath/#Rel) be useful for this?

keloyang replied:
I'm afraid filepath.Rel don't work for this.
```
        base := "tmp"
	p:="tmp_/a"
	rel, err := filepath.Rel(base, p)
	fmt.Printf("%q: %q %v\n", p, rel, err)
```
output is "../tmp_/a" , not "tmp_/a"

yummypeng replied:
Can this pr fix the following issue?
```
# docker run -tid --name test ubuntu /bin/bash -c "mkdir -p /tmp/111/222/ && cd /tmp/111/222 && echo hi > a && ln a b"
8b92fc25e0266d75f22d556a77022d641b91f8b6e6b817065c1ba6a38e34c40b
# docker cp test:/tmp/111/222 hhh/
link /root/222/a /root/hhh/b: no such file or directory
```

thaJeztah commented on pkg/archive/archive.go:
```diff
@@ -530,7 +530,21 @@ func createTarFile(path, extractDir string, hdr *tar.Header, reader io.Reader, L
+		//hdr.Linkname = "tmp/a"
+		//hdr.Name = "tmp_/b"
+		//--> lpath = tmp_/a -->targetPath =extractDir + lpath --> ln tmp_/a tmp_/b
+		lpath := hdr.Linkname
```
Thinking if all this should be extracted to a function

thaJeztah reviewed (commented):
left some comments/questions 😄

cyphar commented on pkg/archive/archive.go:
```diff
@@ -530,7 +550,8 @@ func createTarFile(path, extractDir string, hdr *tar.Header, reader io.Reader, L
 	case tar.TypeLink:
-		targetPath := filepath.Join(extractDir, hdr.Linkname)
+		lpath := adjustHardLink(hdr.Linkname, hdr.Name)
+		targetPath := filepath.Join(extractDir, lpath)
```
I don't agree that the bug is here. I think the bug is in the generation of the `tar` archive, because it doesn't make sense for `Linkname` to be `tmp/a` in this situation (I actually don't like how we do `docker cp` path fudging but that ship sailed 3 years ago).

jlhawn replied:
I agree that this is the wrong place to make this change. It should probably be done in the `RebaseArchiveEntries` function in the file `pkg/archive/copy.go`. There, it should check to see if `hdr.Typeflag == tar.TypeLink` and replace `hdr.Linkname` in the same way that it replaces `hdr.Name`:

```diff
--- a/pkg/archive/copy.go
+++ b/pkg/archive/copy.go
@@ -332,6 +332,9 @@ func RebaseArchiveEntries(srcContent io.Reader, oldBase, newBase string) io.Read
                        }
 
                        hdr.Name = strings.Replace(hdr.Name, oldBase, newBase, 1)
+                       if hdr.Typeflag == tar.TypeLink {
+                               hdr.Linkname = strings.Replace(hdr.Linkname, oldBase, newBase, 1)
+                       }
 
                        if err = rebasedTar.WriteHeader(hdr); err != nil {
                                w.CloseWithError(err)
```

thaJeztah replied:
Is `Replace()` safe to use here? Looks like it may replace matches in the middle of a path as well (thinking `HasPrefix()` + `TrimPrefix()` to prevent that case)

jlhawn replied:
> ... that ship sailed 3 years ago

It's only been 2 years! https://github.com/moby/moby/pull/13171 🎉 

Also, since the `docker cp` logic that does the name fudging is done entirely in the CLI, we should be able to redesign it without affecting older clients.

One alternative that I have warmed up to is that if the source of the archive is a directory, then create the archive by going into that source directory and archiving `.` such that all archive entries begin with `.` and then the extract logic would be:

1. Create the destination directory if it does not exist.
2. Extract the archive in the destination directory.

That should handle the case where the name of the destination directory is not the same as the source directory. That would do away with the need to change the archive entries on the fly.

jlhawn replied:
> Is `Replace()` safe to use here?

I'm pretty sure that in this code path, `oldBase` is guaranteed to be at the beginning of `hdr.Name` but your suggestion is definitely safer. I may be overlooking something, but it should be fine as long as all of the existing CLI tests around `docker cp` still pass.

keloyang replied:
@cyphar, For docker cp, I think the bug  is in  the generation or Untar, but in some other case，the bug is only in Untar. For  example, we save the  tar data(with hard link) to test.tar.gz on the docker daemon when we execute docker cp, or just use tar command to save to test.tar.gz,to uncompress test.tar.gz, the following code doesn't work,
```
package main

import (
	"os"
	"fmt"
	
	"github.com/docker/docker/pkg/archive"
)

func main() {
	in, _ := os.OpenFile(os.Args[1], os.O_RDWR, 0666)
	srcInfo := archive.CopyInfo{
		Path:       "/tmp",
		Exists:     true,
		IsDir:      true,
	}
	if err := archive.CopyTo(in, srcInfo, os.Args[2]); err != nil{
		fmt.Printf("%v\n", err)
	}
}
```
```
root@ubuntu:~/root/tar# go build cp.go
root@ubuntu:~/root/tar# cd tmp/
root@ubuntu:~/root/tar# echo "aaaa">a
root@ubuntu:~/root/tar# ln a b
root@ubuntu:~/root/tar# cd ../
root@ubuntu:~/root/tar# tar xvf test.tar.gz tmp
tmp/
tmp/b
tmp/a

./cp test.tar.gz tmp__
```
after this pr, the code above will work, and I think the change is forward compatible.

cyphar replied:
@jlhawn 

> One alternative that I have warmed up to is that if the source of the archive is a directory, then create the archive by going into that source directory and archiving . such that all archive entries begin with . and then the extract logic would be:

Yup, that's exactly how I would want to do it. Effectively, make the target `filepath.Dir(target)` if it's a file and just `target` if it's a directory.

> I may be overlooking something, but it should be fine as long as all of the existing CLI tests around docker cp still pass.

I'm personally a little more cautious (given how many different breakouts I've found with `docker cp` in the past). If we can do it with `filepath` we should.

cyphar replied:
@keloyang I'm not sure I understand what you're saying. The problem is that when you generate the archive the `hdr.Linkname`s are not properly translated to the target path (they are correctly computed in the source but not in the destination). If we fix the archive generation there's no need to change clients, and you avoid having this on-the-fly rewriting (that makes me quite nervous).

keloyang replied:
@cyphar,I misunderstood it before,thanks for interpretation.

keloyang commented:
@jlhawn thanks, your code can also work and I think it is better for docker cp.

I think we still need add extractDir != "." ,PTAL.
```
-               if !strings.HasPrefix(targetPath, extractDir) {
+               if !strings.HasPrefix(targetPath, extractDir) && extractDir != "." {
```

cyphar commented:
We shouldn't be using `strings.HasPrefix` for `breakoutError` we should be using something like this but with better handling for `.` (which is what I use in umoci):

```go
// isParent returns whether the path a is lexically an ancestor of the path b.
func isParent(a, b string) bool {
	a = filepath.Clean(a)
	b = filepath.Clean(b)

	for a != b && b != filepath.Dir(b) {
		b = filepath.Dir(b)
	}
	return a == b
}
```

I'll submit a PR to clean up all of these cases.

Commit:
Fix docker cp dir with hard link

docker run --name=test ubuntu /bin/sh -c "cd /tmp && echo hi > a && ln a b" && docker cp test:/tmp tmp_
test
link /root/tmp/a /root/tmp_/b: no such file or directory

Signed-off-by: yangshukui <yangshukui@huawei.com>

jlhawn commented:
> Do you know why we're not using FollowSymlinkInScope in pkg/archive?

I'm not sure, no.

keloyang commented:
ping @thaJeztah

thaJeztah reviewed (approved):
LGTM, sorry for the delay :blush:

tonistiigi commented:
LGTM

thaJeztah commented:
Thanks @keloyang!

@cyphar agree with your follow-up PR 👍
